nil/services/rollup/l1_fetcher.go

```diff
@@ -0,0 +1,114 @@
+package rollup
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/NilFoundation/nil/nil/common/concurrent"
+	"github.com/NilFoundation/nil/nil/common/logging"
+	l1types "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rs/zerolog"
+)
+
+//go:generate go run github.com/matryer/moq -out l1_fetcher_generated_mock.go -rm -stub -with-resets . L1BlockFetcher
+
+const pollInterval = 5 * time.Second
+
+type L1BlockFetcher interface {
+	GetLastBlockInfo(ctx context.Context) (*l1types.Header, error)
+}
+
+type L1BlockFetcherRpc struct {
+	ctx       context.Context
+	client    *rpc.Client
+	header    *l1types.Header
+	logger    zerolog.Logger
+	nodeIndex int
+}
+
+var rpcNodeList = []string{
+	"https://eth.llamarpc.com",
+	"https://eth-mainnet.public.blastapi.io",
+	"https://rpc.ankr.com/eth",
+	"https://eth-mainnet.public.blastapi.io",
+	"https://rpc.flashbots.net",
+	"https://cloudflare-eth.com",
+}
+
+func NewL1BlockFetcherRpc(ctx context.Context) L1BlockFetcher {
+	p := &L1BlockFetcherRpc{
+		ctx:    ctx,
```

Comment:
I think it's better to pass context directly to each function. You don't have any complex callback systems to keep context as a struct field.